tests/test_regularization.py

```diff
@@ -0,0 +1,436 @@
+        # within the tolerable interval
+        val_loss = train_loss * (tolerable_interval[0] + tolerable_interval[1]) / 2
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+    else:
+        # we have a zero loss. We try two cases. When the validation loss is zero,
+        # the ratio is undefined, so we should return the current lambda. When the
+        # validation loss is nonzero, the ratio is infinite, so we should see the lambda
+        # increase by the scaling factor.
+        # We try it for both a tensor and a float value.
+        val_loss = th.tensor(0.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = 0.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_
+        val_loss = th.tensor(1.0)
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+        val_loss = 1.0
+        assert scaler(lambda_, train_loss, val_loss) == lambda_ * (1 + scaling_factor)
+
+
+def test_interval_param_scaler_raises(interval_param_scaler):
+    scaler = interval_param_scaler
+    with pytest.raises(ValueError, match="val_loss and train_loss must be scalars"):
+        scaler(0.0, th.Tensor([1.0, 2.0]), th.Tensor([3.0, 4.0]))
+    with pytest.raises(ValueError, match="val_loss and train_loss must be scalars"):
+        scaler(0.0, "random value", "random value")  # type: ignore
+    with pytest.raises(ValueError, match="lambda_ must be a float"):
+        scaler(th.tensor(1.0), 1.0, 1.0)  # type: ignore
+    with pytest.raises(ValueError, match="lambda_ must not be zero.*"):
+        scaler(0.0, 1.0, 1.0)
+
+
+def test_interval_param_scaler_init_raises():
+    # this validates the value of scaling_factor.
+    with pytest.raises(
+        ValueError,
+        match=r"scaling_factor must be in \(0, 1\) within machine precision.",
+    ):
+        # cannot be negative as this is counter-intuitive to
+        # the direction of scaling (just use the reciprocal).
+        updaters.IntervalParamScaler(-1, (0.1, 0.9))
+    with pytest.raises(
+        ValueError,
+        match=r"scaling_factor must be in \(0, 1\) within machine precision.",
+    ):
+        # cannot be larger than one as this would make lambda
+        # negative when scaling down.
+        updaters.IntervalParamScaler(1.1, (0.1, 0.9))
+    with pytest.raises(
+        ValueError,
+        match=r"scaling_factor must be in \(0, 1\) within machine precision.",
+    ):
+        # cannot be exactly zero, as this never changes the value
+        # of lambda when scaling up.
+        updaters.IntervalParamScaler(0.0, (0.1, 0.9))
+    with pytest.raises(
+        ValueError,
+        match=r"scaling_factor must be in \(0, 1\) within machine precision.",
+    ):
+        # cannot be exactly one, as when lambda is scaled down
+        # this brings it to zero.
+        updaters.IntervalParamScaler(1.0, (0.1, 0.9))
+
+    # an interval obviously needs two elements only.
+    with pytest.raises(
```

Comment:
You might enjoy the library `hypothesis` for testing things like this. Tests as written are fine, just might save you time in future for checking invariants like this.

Reply:
Haha okay. I just mentioned it on a comment above, seems like you already know about it. I don't think it's worth re-writing now, but will certainly considering using it in the future.